weed/s3api/s3api_acp.go

```diff
@@ -1,28 +1,329 @@
+	if accountId == objectOwner {
+		return bucketMetadata, objectEntry, objectOwner, s3err.ErrNone
+	}
+
+	objectGrants := s3acl.GetAcpGrants(objectEntry.Extended)
+	if objectGrants != nil {
+		for _, objectGrant := range objectGrants {
+			for _, requiredGrant := range reqGrants {
+				if reflect.DeepEqual(objectGrant, requiredGrant) {
+					return bucketMetadata, objectEntry, objectOwner, s3err.ErrNone
+				}
+			}
+		}
+	}
+
+	glog.V(3).Infof("acl denied! request account id: %s", accountId)
+	return nil, nil, "", s3err.ErrAccessDenied
+}
+
+// Check Object-Read related access
+// includes:
+// - GetObjectHandler
+//
+// offload object access validation to Filer layer
+// - s3acl.CheckObjectAccessForReadObject
+func (s3a *S3ApiServer) checkBucketAccessForReadObject(r *http.Request, bucket, object string) s3err.ErrorCode {
+	bucketMetadata, errCode := s3a.bucketRegistry.GetBucketMetadata(bucket)
+	if errCode != s3err.ErrNone {
+		return errCode
+	}
+
+	if bucketMetadata.ObjectOwnership != s3_constants.OwnershipBucketOwnerEnforced {
+		//offload object acl validation to filer layer
+		r.Header.Set(s3_constants.X_SeaweedFS_Header_Amz_Bucket_Owner_Id, *bucketMetadata.Owner.ID)
+	}
+
+	return s3err.ErrNone
+}
+
+//Check ObjectAcl-Read related access
+// includes:
+// - GetObjectAclHandler
+func (s3a *S3ApiServer) checkAccessForReadObjectAcl(r *http.Request, bucket, object string) (acp *s3.AccessControlPolicy, errCode s3err.ErrorCode) {
+	bucketMetadata, errCode := s3a.bucketRegistry.GetBucketMetadata(bucket)
+	if errCode != s3err.ErrNone {
+		return nil, errCode
+	}
+
+	getAcpFunc := func() (*s3.AccessControlPolicy, s3err.ErrorCode) {
+		entry, err := getObjectEntry(s3a, bucket, object)
+		if err != nil {
+			if err == filer_pb.ErrNotFound {
+				return nil, s3err.ErrNoSuchKey
+			} else {
+				return nil, s3err.ErrInternalError
+			}
+		}
+
+		if entry.IsDirectory {
+			return nil, s3err.ErrExistingObjectIsDirectory
+		}
+
+		acpOwnerId := s3acl.GetAcpOwner(entry.Extended, *bucketMetadata.Owner.ID)
+		acpOwnerName := s3a.accountManager.IdNameMapping[acpOwnerId]
+		acpGrants := s3acl.GetAcpGrants(entry.Extended)
+		acp = &s3.AccessControlPolicy{
+			Owner: &s3.Owner{
+				ID:          &acpOwnerId,
+				DisplayName: &acpOwnerName,
+			},
+			Grants: acpGrants,
+		}
+		return acp, s3err.ErrNone
+	}
+
+	if bucketMetadata.ObjectOwnership == s3_constants.OwnershipBucketOwnerEnforced {
+		return getAcpFunc()
+	} else {
+		accountId := s3acl.GetAccountId(r)
+
+		acp, errCode := getAcpFunc()
+		if errCode != s3err.ErrNone {
+			return nil, errCode
+		}
+
+		if accountId == *acp.Owner.ID {
+			return acp, s3err.ErrNone
+		}
+
+		//find in Grants
+		if acp.Grants != nil {
+			reqGrants := s3acl.DetermineReqGrants(accountId, s3_constants.PermissionReadAcp)
+			for _, requiredGrant := range reqGrants {
+				for _, grant := range acp.Grants {
+					if reflect.DeepEqual(requiredGrant, grant) {
```

Comment:
Try not to use reflection. Later there might be a need to customize the comparison.

Reply:
replaced with `weed/s3api/s3acl/acl.go:437`